File: java/base/src/main/java/vsdk/toolkit/common/dataStructures/CircularListBaseIterator.java
package vsdk.toolkit.common.dataStructures;

public class CircularListBaseIterator {
    private CircularListLink currentElement;
    private CircularListBase currentList;

    public CircularListBaseIterator(CircularListBase list) {
        init(list);
    }

    public void init(CircularListBase list) {
        currentList = list;
        currentElement = currentList.lastLink();
    }

    public CircularListLink next() {
        CircularListLink response;

        if (currentElement == null) {
            response = null;
        }
        else {
            currentElement = currentElement.nextLink;
            response = currentElement;
        }

        if (currentElement == currentList.lastLink()) {
            currentElement = null;
        }

        return response;
    }
}

File: cpp/base/src/main/vsdk/toolkit/common/dataStructures/CircularListBaseIterator.cpp
#include "vsdk/toolkit/common/dataStructures/CircularListBaseIterator.h"

CircularListBaseIterator::CircularListBaseIterator(CircularListBase &list): currentElement(), currentList() {
    init(list);
}

CircularListBaseIterator::~CircularListBaseIterator() {
}

void
CircularListBaseIterator::init(CircularListBase &list) {
    currentList = &list;
    currentElement = currentList->lastLink();
}

CircularListLink *
CircularListBaseIterator::next() {
    CircularListLink *response;

    if ( currentElement == nullptr ) {
        response = nullptr;
    } else {
        currentElement = currentElement->nextLink;
        response = currentElement;
    }

    if ( currentElement == currentList->lastLink() ) {
        currentElement = nullptr;
    }

    return response;
}
